timefold-solver/src/main/java/com/baeldung/timefoldsolver/ShiftScheduleConstraintProvider.java

```diff
@@ -0,0 +1,37 @@
+package com.baeldung.timefoldsolver;
+
+import ai.timefold.solver.core.api.score.buildin.hardsoft.HardSoftScore;
+import ai.timefold.solver.core.api.score.stream.Constraint;
+import ai.timefold.solver.core.api.score.stream.ConstraintFactory;
+import ai.timefold.solver.core.api.score.stream.ConstraintProvider;
+
+import static ai.timefold.solver.core.api.score.stream.Joiners.equal;
+
+public class ShiftScheduleConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
```

Comment:
Some feedback on the ConstraintProvider API, it's unusual to use array types in a modern library API, I would have expected a List<Constraint> instead. But I appreciate this is something that is not an easy thing to change if you have an existing userbase.

Reply:
Thanks for the feedback.
It's intentional design, at least at the time, because the array of constraints never changes size. And now we're stuck with it for Backwards Compatibility for now.

It looks old indeed :/